compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3135,34 +3140,30 @@ generateS390CompareAndBranchOpsHelper(TR::Node * node, TR::CodeGenerator * cg, T
             cg->evaluate(secondChild);
          }
 
-      switch (childType)
+      switch (TR::DataType::getSize(firstChild->getDataType()))
          {
-         case TR::Int64:
-         case TR::Address:
+         case 8:
             temp.genericAnalyser (node, isUnsignedCmp ? TR::InstOpCode::CLGR : TR::InstOpCode::CGR,
-                                 isUnsignedCmp ? TR::InstOpCode::CLG : TR::InstOpCode::CG, TR::InstOpCode::LGR,
-                                 true, branchTarget, fBranchOpCond, rBranchOpCond);
+                                 isUnsignedCmp ? TR::InstOpCode::CLG : TR::InstOpCode::CG,
+                                 TR::InstOpCode::LGR, true, branchTarget, fBranchOpCond, rBranchOpCond);
             returnInstruction = cg->getAppendInstruction();
             isBranchGenerated = true;
             break;
-         case TR::Int32:
-            TR_ASSERT( dataType == TR::Int32 || comp->useCompressedPointers(), "First child, %p, and second child, %p, data types do not match. This is an issue when compression is disabled.", firstChild, secondChild);
+         case 4:
+         case 1:
```

Comment:
This looks incorrect for a byte operand. The register to memory compare will use a `C` instruction which is not correct. On Z we don't have an RX type instruction to compare with a byte in memory. We cannot handle this case. The code before this change was incorrect as well. Can you remove this case and turn the assert into a fatal assert instead in the `default` case, and modify the assert to print the data type coming in as well as the size?